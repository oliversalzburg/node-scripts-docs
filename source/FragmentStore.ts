import fs from "fs/promises";
import { Tokens, marked } from "marked";
import path from "path";
import { DOCUMENTATION_PENDING_DEFAULT } from "./FragmentRenderer";

export const DOCS_FRAGMENTS_DEFAULT_LOCATION = "docs/reference/Repository Scripts/";

export class DocumentationFragment {
  descriptionMarkdown: string;
  filename: string;

  constructor(filename: string, descriptionMarkdown = DOCUMENTATION_PENDING_DEFAULT) {
    this.filename = filename;
    this.descriptionMarkdown = descriptionMarkdown;
  }
}

export class FragmentStore {
  private fragmentDirectory: string;

  public fragments = new Map<string, DocumentationFragment>();

  constructor(fragmentDirectory: string) {
    this.fragmentDirectory = fragmentDirectory;
  }

  async loadFragment(fragmentFilename: string) {
    const assumedScriptName = FragmentStore.fragmentFilenameToScript(fragmentFilename);
    if (!assumedScriptName) {
      throw new Error(`Unable to interpret fragment file name '${fragmentFilename}'!`);
    }

    const existingDocumentation = await fs.readFile(
      path.resolve(this.fragmentDirectory, fragmentFilename),
      "utf-8",
    );
    const parsed = marked.lexer(existingDocumentation);

    // Expect parsed to be an array of [heading, list]
    const items = (parsed?.[1] as Tokens.List)?.items;
    // Expect items to be [list_item, list_item, list_item]
    const descriptionItem =
      items &&
      items.find(
        item => 1 <= item.tokens.length && (item.tokens[0] as Tokens.Text).text === "Description:",
      );
    if (!descriptionItem) {
      throw new Error(
        `Unable to find description item in documentation fragment at '${fragmentFilename}'!`,
      );
    }

    const descriptionRaw = descriptionItem.tokens
      .slice(2)
      .map(token => token.raw)
      .join("");

    this.fragments.set(
      assumedScriptName,
      new DocumentationFragment(fragmentFilename, descriptionRaw),
    );
  }

  getExistingDescription(scriptName: string) {
    if (!this.fragments.has(scriptName)) {
      return undefined;
    }

    return this.fragments.get(scriptName)!.descriptionMarkdown;
  }

  static scriptToFragmentFilename(scriptName: string) {
    // $ needs to be duplicated, as it's otherwise interpreted as part of $1,$2,$3,... references
    return `.${scriptName.replace(/\:/g, "$$$$")}.md`;
  }

  static fragmentFilenameToScript(fragmentName: string) {
    const matches = /\.(.+?)\.md/.exec(fragmentName);
    if (!matches) {
      return null;
    }

    return matches[1].replace(/\$\$/g, ":");
  }

  static async exists(storePath: string) {
    try {
      await fs.stat(storePath);
    } catch {
      return false;
    }

    return true;
  }
}
