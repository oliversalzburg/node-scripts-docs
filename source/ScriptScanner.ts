import { globby } from "globby";
import fs from "node:fs/promises";
import path, { join } from "node:path";
import { ScriptStore } from "./ScriptStore.js";

/**
 * The important parts of a project's manifest.
 */
export interface Manifest {
  /**
   * The name of the project.
   */
  name: string;

  /**
   * The version of the project.
   */
  version?: string | undefined;

  /**
   * The scripts that exist in the manifest.
   */
  scripts?: Record<string, string> | undefined;
}

/**
 * Searches for scripts in a project.
 */
export class ScriptScanner {
  private rootDirectory: string;
  private workspaces: Array<string>;

  /**
   * Manifests we found in the project.
   */
  public manifests = new Array<string>();

  /**
   * Constructs a new script scanner.
   * @param rootDirectory - The root directory to search for manifests.
   * @param workspaces - The location where workspaces are stored in the monorepo.
   */
  constructor(rootDirectory: string, workspaces = ["packages/*"]) {
    this.rootDirectory = rootDirectory;
    this.workspaces = workspaces;
  }

  /**
   * Find manifests in the target project.
   * @param includeRootManifest - Should the manifest of the project root be included?
   * @returns The found manifests.
   */
  async findManifests(includeRootManifest = true) {
    const manifests = new Array<string>();
    for (const workspacePath of this.workspaces) {
      const workspaceManifests = await globby(join(workspacePath, "package.json"), {
        cwd: this.rootDirectory,
      });
      manifests.push(...workspaceManifests);
    }

    if (includeRootManifest) {
      manifests.push("package.json");
    }

    return manifests;
  }

  /**
   * Find manifests and load them into the scanner instance.
   */
  async loadManifests() {
    this.manifests = await this.findManifests();
  }

  /**
   * Find all scripts in all manifests and put them into a script store.
   * @returns An instance of a script store.
   */
  async loadScripts() {
    const scriptStore = new ScriptStore(this.rootDirectory);
    for (const manifestPath of this.manifests) {
      const isRootManifest = manifestPath === "package.json";
      const manifest = await ScriptScanner.loadManifest(
        path.resolve(this.rootDirectory, manifestPath),
      );
      const projectName = manifest.name;
      const projectScripts: Record<string, string> = manifest.scripts ?? {};

      for (const [name, script] of Object.entries(projectScripts)) {
        const isGlobal = name.includes(":");
        scriptStore.add(manifestPath, projectName, name, script, isGlobal, isRootManifest);
      }
    }

    return scriptStore;
  }

  /**
   * Loads the target manifest.
   * @param manifestPath - The path of the manifest.
   * @returns The loaded manifest.
   */
  static async loadManifest(manifestPath: string) {
    const manifestContent = await fs.readFile(manifestPath, "utf-8");
    const manifest = JSON.parse(manifestContent) as Manifest;
    return manifest;
  }

  /**
   * Check if the target location can be loaded.
   * @param manifestPath - The location to check.
   * @returns `true` if the manifest exists; `false` otherwise.
   */
  static async canLoad(manifestPath: string) {
    try {
      await fs.stat(manifestPath);
    } catch {
      return false;
    }

    return true;
  }
}
