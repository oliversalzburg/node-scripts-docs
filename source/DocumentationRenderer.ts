import fs from "node:fs/promises";
import path from "node:path";
import { makeDocumentation } from "./FragmentRenderer.js";
import { FragmentStore } from "./FragmentStore.js";
import { ScriptStore } from "./ScriptStore.js";

export class DocumentationRenderer {
  private metadata: ScriptStore;

  constructor(metadata: ScriptStore) {
    this.metadata = metadata;
  }

  render(includeLocalScripts = false) {
    let fullDocument = "# Full reference\n\n";
    for (const scriptMeta of this.metadata.scripts) {
      if (!scriptMeta.isGlobal && !scriptMeta.isRootManifest && !includeLocalScripts) {
        continue;
      }

      const scriptDocs = makeDocumentation(scriptMeta, scriptMeta.description);
      fullDocument += scriptDocs + "\n";
    }
    return fullDocument;
  }

  async flushFragments(rootDirectory: string, includeLocalScripts = false) {
    await fs.mkdir(rootDirectory, { recursive: true });
    for (const scriptMeta of this.metadata.scripts) {
      if (!scriptMeta.isGlobal && !scriptMeta.isRootManifest && !includeLocalScripts) {
        continue;
      }

      //const description = fragmentStore.getExistingDescription(scriptMeta.scriptName);

      const scriptDocs = makeDocumentation(scriptMeta, scriptMeta.description);
      const fragmentFilename = FragmentStore.scriptToFragmentFilename(scriptMeta.scriptName);
      const fragmentPath = path.resolve(rootDirectory, fragmentFilename);
      await fs.writeFile(fragmentPath, scriptDocs);
    }
  }
}
