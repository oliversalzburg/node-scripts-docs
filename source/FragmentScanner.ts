import { globby } from "globby";
import { FragmentStore } from "./FragmentStore.js";

/**
 * Loads any fragments from the given directory and places them into a fragment store.
 * @param fragmentDirectory - The directory from which to load fragments.
 * @returns A fragment store.
 */
export const loadFragments = async (fragmentDirectory: string) => {
  const fragmentStore = new FragmentStore(fragmentDirectory);
  const fragments = await globby(FragmentStore.scriptToFragmentFilename("*"), {
    cwd: fragmentDirectory,
  });
  for (const fragment of fragments) {
    await fragmentStore.loadFragment(fragment);
  }

  return fragmentStore;
};
