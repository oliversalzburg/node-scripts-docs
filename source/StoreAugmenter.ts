import { isDefaultDescription } from "./FragmentRenderer.js";
import { FragmentStore } from "./FragmentStore.js";
import { ScriptStore } from "./ScriptStore.js";

export class StoreAugmenter {
  scriptStore: ScriptStore;

  constructor(scriptStore: ScriptStore) {
    this.scriptStore = scriptStore;
  }

  augment(fragmentStore: FragmentStore) {
    for (const scriptMeta of this.scriptStore.scripts) {
      const description = fragmentStore.getExistingDescription(scriptMeta.scriptName);
      if (description && !isDefaultDescription(description)) {
        scriptMeta.description = description;
      }
    }
  }
}
