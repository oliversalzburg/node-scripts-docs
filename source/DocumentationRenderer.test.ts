import path from "node:path";
import { DocumentationRenderer } from "./DocumentationRenderer.js";
import { loadFragments } from "./FragmentScanner.js";
import { DOCS_FRAGMENTS_DEFAULT_LOCATION } from "./FragmentStore.js";
import { ScriptScanner } from "./ScriptScanner.js";
import { StoreAugmenter } from "./StoreAugmenter.js";

it("renders a script store as expected", async () => {
  const storeFragments = await loadFragments(
    path.join("test/fixtures/default", DOCS_FRAGMENTS_DEFAULT_LOCATION),
  );

  const scannerScripts = new ScriptScanner("test/fixtures/default");
  await scannerScripts.loadManifests();
  const storeScripts = await scannerScripts.loadScripts();

  const augmenter = new StoreAugmenter(storeScripts);
  augmenter.augment(storeFragments);

  const renderer = new DocumentationRenderer(storeScripts);
  expect(renderer.render()).toMatchSnapshot();
  expect(renderer.render(true)).toMatchSnapshot();
});
