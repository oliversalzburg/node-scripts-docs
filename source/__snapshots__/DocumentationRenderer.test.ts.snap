// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`renders a script store as expected 1`] = `
"# Full reference

## build

-   Project: \`default\`
-   Source:

    \`\`\`shell
    tsc
    \`\`\`

-   Description:

    Build the TypeScript sources.

## docs:build

-   Project: \`default\`
-   Source:

    \`\`\`shell
    docker run --rm -it -p 8000:8000 -v \${PWD}:/docs squidfunk/mkdocs-material build --site-dir=public
    \`\`\`

-   Description:

    Build the documentation website locally.

## docs:scripts

-   Project: \`default\`
-   Source:

    \`\`\`shell
    yarn nsd-debug --include-locals
    \`\`\`

-   Description:

    Rebuilds the \`node-scripts-docs\` documentation for this project itself.

## docs:serve

-   Project: \`default\`
-   Source:

    \`\`\`shell
    docker run --rm -it -p 8000:8000 -v \${PWD}:/docs squidfunk/mkdocs-material
    \`\`\`

-   Description:

    Runs the mkdocs-material development server.

## npm:publish

-   Project: \`default\`
-   Source:

    \`\`\`shell
    yarn build && npm publish
    \`\`\`

-   Description:

    Build the latest sources and then \`npm publish\` the current state.

## npm:publish:patch

-   Project: \`default\`
-   Source:

    \`\`\`shell
    yarn build && npm version patch && npm publish
    \`\`\`

-   Description:

    _documentation pending_

## nsd-debug

-   Project: \`default\`
-   Source:

    \`\`\`shell
    yarn build && node output/node-scripts-docs.js
    \`\`\`

-   Description:

    Build the latest sources and then use the build output to execute your command. Parameters are passed through by \`yarn\`.

## test

-   Project: \`default\`
-   Source:

    \`\`\`shell
    jest
    \`\`\`

-   Description:

    _documentation pending_

## test:coverage

-   Project: \`default\`
-   Source:

    \`\`\`shell
    jest --coverage
    \`\`\`

-   Description:

    _documentation pending_

"
`;

exports[`renders a script store as expected 2`] = `
"# Full reference

## build

-   Project: \`default\`
-   Source:

    \`\`\`shell
    tsc
    \`\`\`

-   Description:

    Build the TypeScript sources.

## docs:build

-   Project: \`default\`
-   Source:

    \`\`\`shell
    docker run --rm -it -p 8000:8000 -v \${PWD}:/docs squidfunk/mkdocs-material build --site-dir=public
    \`\`\`

-   Description:

    Build the documentation website locally.

## docs:scripts

-   Project: \`default\`
-   Source:

    \`\`\`shell
    yarn nsd-debug --include-locals
    \`\`\`

-   Description:

    Rebuilds the \`node-scripts-docs\` documentation for this project itself.

## docs:serve

-   Project: \`default\`
-   Source:

    \`\`\`shell
    docker run --rm -it -p 8000:8000 -v \${PWD}:/docs squidfunk/mkdocs-material
    \`\`\`

-   Description:

    Runs the mkdocs-material development server.

## npm:publish

-   Project: \`default\`
-   Source:

    \`\`\`shell
    yarn build && npm publish
    \`\`\`

-   Description:

    Build the latest sources and then \`npm publish\` the current state.

## npm:publish:patch

-   Project: \`default\`
-   Source:

    \`\`\`shell
    yarn build && npm version patch && npm publish
    \`\`\`

-   Description:

    _documentation pending_

## nsd-debug

-   Project: \`default\`
-   Source:

    \`\`\`shell
    yarn build && node output/node-scripts-docs.js
    \`\`\`

-   Description:

    Build the latest sources and then use the build output to execute your command. Parameters are passed through by \`yarn\`.

## test

-   Project: \`default\`
-   Source:

    \`\`\`shell
    jest
    \`\`\`

-   Description:

    _documentation pending_

## test:coverage

-   Project: \`default\`
-   Source:

    \`\`\`shell
    jest --coverage
    \`\`\`

-   Description:

    _documentation pending_

"
`;
